### DISCUSSION TITLE
HTML entities get converted to characters in translations

### DISCUSSION THREAD

**Discussion by renestach** [WeblateOrg/weblate#6657] [Q&A]
Hi!

We have source strings like this:
`<string name="example">This is &lt;b&gt;bold&lt;/b&gt; text.</string>`

Whenever a string like this is translated, it gets converted to 
`<string name="example">This is <b>bold</b> text.</string>`

We have tried to activate safe-html as flag and enforced check but still the same is happening. Is there any way to make Weblate save those HTML entities again and not convert them into characters? If our approach is bad, we are happy for any suggestions.

René

**Comment by nijel** [comment]
This is how the Android strings behave in the translate-toolkit (the library Weblate uses for parsing translation files) - once the string is a valid XML, it is inserted as XML.

**Reply by renestach** [reply]
Thank you for your feedback. It is a bit annoying that our resource files get modified that way but if there is no way to get that behaviour configured, we will have to teach our systems to live with that.

Just for the records: It also seems that line breaks and white spaces get removed. When we add nicely formatted HTML code with line breaks these get removed in all translations.

**Source string**
`<string name="example"><p>Hello</p>`
`    <ul>`
`     <li>First bullet</li>`
`    </ul>`
`</string>`

**Translated string**
`<string name="example"><p>Hello</p> <ul> <li>First bullet</li> </ul></string>`

**Reply by nijel** [reply]
This is the tricky part of any XML format - when you get a string to serialize, it is not clear whether it should be treated as XML or not. The only proper solution is to have a UI toggle for this, but we've done some testing in this and that doesn't work well either - in most people translators do not understand the consequences here.

So, what we have now is:

* XLIFF is always treated as XML, what is not liked by some users, see https://github.com/WeblateOrg/weblate/issues/3081
* Android does autodetection, and it causes issues like this one

**Comment by renestach** [comment]
Hello! I would like to ask the same question again to check whether a solution is available now.

The situation is still the same. We have strings in the XML file that contain HTML entities. These are displayed as the characters they represent in the translation web interface (probably done by the browser), and when the string is saved, it contains the special characters instead of the HTML entities. As a result, less-than and greater-than characters end up in the XML file, which makes it invalid and causes issues in our Android and iOS apps.

Example:

Our source strings look like this:  
`<string name="example">This is &lt;b&gt;bold&lt;/b&gt; text.</string>`

In the weblate interface this string is displayed as 
`This is <b>bold</b> text.`

And finally the string is saved back to the XML file as
`<string name="example">This is <b>bold</b> text.</string>`

Is there any way to prevent this from happening or to fix it before the file is committed?

**Reply by nijel** [reply]
https://github.com/WeblateOrg/weblate/issues/11995 is tracking that. I hope to make it for the upcoming 5.17 release.

**Reply by renestach** [reply]
That is good news. What kind of solution do you have in mind?

To be more specific, I think the best way to solve this is to make sure that XML files are always saved with valid XML content by automatically converting required characters into HTML entities. This would also fix translation files that are already broken. To avoid surprising users, this feature could be made optional.
